lci_strategic_plugin/src/lci_strategic_plugin.cpp

```diff
@@ -0,0 +1,849 @@
+                                            const lanelet::CarmaTrafficSignalPtr& traffic_light, const lanelet::ConstLanelet& entry_lanelet, const lanelet::ConstLanelet& exit_lanelet, const lanelet::ConstLanelet& current_lanelet)
+{
+  if (!traffic_light)  // If we are in the approaching state and there is no longer any lights ahead of us then
+                       // the vehicle must have crossed the stop bar
+  {
+    transition_table_.signal(TransitEvent::CROSSED_STOP_BAR);
+    return;
+  }
+
+  auto stop_line = traffic_light->getStopLine(entry_lanelet);
+
+  if (!stop_line)
+  {
+    throw std::invalid_argument("Given entry lanelet doesn't have stop_line...");
+  }
+
+  double traffic_light_down_track =
+      wm_->routeTrackPos(stop_line.get().front().basicPoint2d()).downtrack;
+
+  ROS_DEBUG("traffic_light_down_track %f", traffic_light_down_track);
+
+  double distance_remaining_to_traffic_light = traffic_light_down_track - current_state.downtrack;
+
+  // If the vehicle is at a stop trigger the stopped state
+  constexpr double HALF_MPH_IN_MPS = 0.22352;
+  if (current_state.speed < HALF_MPH_IN_MPS &&
+      fabs(distance_remaining_to_traffic_light) < config_.stopping_location_buffer)
+  {
+    transition_table_.signal(TransitEvent::STOPPED);  // The vehicle has come to a stop at the light
+    return;
+  }
+
+  // At this point we know the vehicle is within the activation distance and we know the current and next light phases
+  // All we need to now determine is if we should stop or if we should continue
+
+  intersection_speed_ = findSpeedLimit(entry_lanelet); //technically lanelet is not "inside the lanelet"
+
+  double speed_limit = findSpeedLimit(current_lanelet);
+
+  ROS_DEBUG_STREAM("intersection_speed_: " << intersection_speed_.get());
+
+  intersection_end_downtrack_ =
+      wm_->routeTrackPos(exit_lanelet.centerline2d().front().basicPoint2d()).downtrack;
+
+  ROS_DEBUG_STREAM("intersection_end_downtrack_: " << intersection_end_downtrack_.get());
+
+  // Start of TSMO UC2 Algorithm
+
+  ros::Time earliest_entry_time = current_state.stamp + get_earliest_entry_time(distance_remaining_to_traffic_light, speed_limit, 
+                                                  current_state.speed, intersection_speed_.get(), max_comfort_accel_, max_comfort_decel_);
+
+  ROS_DEBUG_STREAM("earliest_entry_time: " << std::to_string(earliest_entry_time.toSec()) << ", with : " << earliest_entry_time - current_state.stamp  << " left at: " << std::to_string(current_state.stamp.toSec()));
+
+  ros::Time nearest_green_entry_time = get_nearest_green_entry_time(current_state.stamp, earliest_entry_time, traffic_light) 
+                                          + ros::Duration(config_.green_light_time_buffer + 0.2); //0.2sec more buffer since green_light buffer also ends at previous state
+                                                                                                  // TODO need review for above
```

Comment:
it is the addition of green_buffer as well as 0.2 number, which is in seconds. 
- the get_nearest_green_time algorithm can return exactly the start of the green phase. So we would always miss green if we account for lateness. So I was wondering if it is correct to add green_buffer to the algorithm's result?
- adding 0.2s is also to avoid the predictState getting non-green due to late arrival time exactly falling on the end_time. is 0.2s good amount?

Reply:
The usage of the buffer here I think might be a bit problematic since this value is feeding into the rest of the algorithm, seems like it could skew your results. 
I think a buffer would be what you apply at the final check so for example in your canArriveAtGreenWithCertainty method or some such location. For initial integration testing though, it might be best to set the buffer to zero to see if your initial plans are giving the correct result.  Then you can increase the buffer to account for the inaccuracy in vehicle prediction. 


As far as the 0.2s goes, if the only purpose is to prevent exact time matching then I would use and even smaller value like 0.01s. A yellow phase might be only 2s long which makes 0.2 about 10% of the phase time which is pretty substantial for an epsilon value.

Reply:
Okay I have set the buffer to 0 and reduced 0.2s to 0.01s for the moment.